src/modules/roc_address/io_uri_format.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_address/io_uri.h"
+#include "roc_address/pct.h"
+#include "roc_core/panic.h"
+
+namespace roc {
+namespace address {
+
+namespace {
+
+bool str_append(char* buf, size_t bufsz, size_t& pos, const char* str) {
+    const size_t orig_len = strlen(str);
+
+    size_t len = orig_len;
+    if (len > bufsz - pos - 1) {
+        len = bufsz - pos - 1;
+    }
+
+    if (len > 0) {
+        memcpy(buf + pos, str, len);
+        pos += len;
+        buf[pos] = '\0';
+    }
+
+    return (len == orig_len);
+}
+
+} // namespace
+
+bool format_io_uri(const IoURI& u, char* buf, size_t buf_size) {
+    roc_panic_if(buf == NULL);
```

Comment:
I followed the following logic.

The first case is an incorrect usage and indicates some serious problem (some code is working with a buffer but there is no buffer), so it's better to panic.

The second case may be, arguably, a correct usage, or at least the code causing this problem may look quite correct :) For example, the code is incrementally appending to the buffer and subtracting from the buffer size, and at some point the buffer size became zero. Such things are happening often enough, so it's better to handle them and return an error instead of a panic.

In general, I'm just replicating contracts of the usual string functions like strncat() or sprintf(). They usually require a non-null buffer but can handle zero buffer/string size.